src/nginx.py

```diff
@@ -297,35 +166,77 @@ def _log_verbose(self, verbose: bool = True) -> List[Dict[str, Any]]:
+                    "distributor"
+                ].union(addresses_by_role[TempoRole.all])
+            else:
+                addresses_mapped_to_upstreams["distributor"] = addresses_by_role[TempoRole.all]
+            if "query_frontend" in addresses_mapped_to_upstreams:
+                addresses_mapped_to_upstreams["query_frontend"] = addresses_mapped_to_upstreams[
+                    "query_frontend"
+                ].union(addresses_by_role[TempoRole.all])
+            else:
+                addresses_mapped_to_upstreams["query_frontend"] = addresses_by_role[TempoRole.all]
+        if "distributor" in addresses_mapped_to_upstreams.keys():
+            nginx_upstreams.extend(
+                self._distributor_upstreams(addresses_mapped_to_upstreams["distributor"])
+            )
+        if "query_frontend" in addresses_mapped_to_upstreams.keys():
+            nginx_upstreams.extend(
+                self._query_frontend_upstreams(addresses_mapped_to_upstreams["query_frontend"])
             )
 
         return nginx_upstreams
 
-    def _locations(self, addresses_by_role: Dict[str, Set[str]]) -> List[Dict[str, Any]]:
-        nginx_locations = LOCATIONS_BASIC.copy()
-        roles = addresses_by_role.keys()
+    def _distributor_upstreams(self, address_set: Set[str]) -> List[Dict[str, Any]]:
+        return [
+            self._upstream("otlp-http", address_set, Tempo.receiver_ports["otlp_http"]),
+            self._upstream("otlp-grpc", address_set, Tempo.receiver_ports["otlp_grpc"]),
+            self._upstream("zipkin", address_set, Tempo.receiver_ports["zipkin"]),
+            self._upstream(
+                "jaeger-thrift-http", address_set, Tempo.receiver_ports["jaeger_thrift_http"]
+            ),
+        ]
+
+    def _query_frontend_upstreams(self, address_set: Set[str]) -> List[Dict[str, Any]]:
+        return [
+            self._upstream("tempo-http", address_set, Tempo.server_ports["tempo_http"]),
+            self._upstream("tempo-grpc", address_set, Tempo.server_ports["tempo_grpc"]),
+        ]
 
-        if "distributor" in roles:
-            nginx_locations.extend(LOCATIONS_DISTRIBUTOR)
-        if "alertmanager" in roles:
-            nginx_locations.extend(LOCATIONS_ALERTMANAGER)
-        if "ruler" in roles:
-            nginx_locations.extend(LOCATIONS_RULER)
-        if "query-frontend" in roles:
-            nginx_locations.extend(LOCATIONS_QUERY_FRONTEND)
-        if "compactor" in roles:
-            nginx_locations.extend(LOCATIONS_COMPACTOR)
+    def _upstream(self, role: str, address_set: Set[str], port: int) -> Dict[str, Any]:
+        return {
+            "directive": "upstream",
+            "args": [role],
+            "block": [{"directive": "server", "args": [f"{addr}:{port}"]} for addr in address_set],
+        }
+
+    def _locations(self, upstream: str, grpc: bool) -> List[Dict[str, Any]]:
+        protocol = "grpc" if grpc else "http"
```

Comment:
shouldn't here we have sth like `protocol = "grpc" if grpc else ("https" if self.tls_ready else "http")`
When we enable TLS, workloads will also be run using TLS. So, internal communication using what's here (http) will fail as its expecting https.

Reply:
True, and also `grpcs` for SSL-enabled setup. Added.